Redesign of the GPIO HIL.

phil-levis opened the pull request:
### Pull Request Overview

This pull request updates the GPIO HIL to more modern Tock programming approaches. It currently includes support for only the imix/SAM4L. Further boards and chips will be added as we become sure about the API.


### Testing Strategy

This code has been compiled and tested on imix, hail, and nrf52. (Others)?


### TODO or Help Wanted

This needs support for the other boards/chips. 


### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

alevy commented on kernel/src/hil/gpio.rs:
```diff
@@ -1,67 +1,268 @@
+    fn fired(&self);
+}
+
+pub trait InterruptWithValue: Input {
```
Trait needs documentation

phil-levis replied:
Added.

alevy commented on kernel/src/hil/gpio.rs:
```diff
@@ -1,67 +1,268 @@
+    fn fired(&self, value: u32);
+}
+
+pub struct InterruptValueWrapper {
```
needs strct comment

phil-levis replied:
Added.

phil-levis commented:
I've merged the stm32f4xx port. This means we currently have sam4l and stm32 support but still need  support for the nrf5x and cc26x2.

alevy commented:
I'll take the nRF5x

niklasad1 commented on kernel/src/hil/gpio.rs:
```diff
@@ -1,67 +1,289 @@
 
-    /// Configure the GPIO pin as an input pin.
-    fn make_input(&self);
+pub trait Configure {
```
trait documentation please,

Also, I find the `Configuration trait` hard to understand with methods such as `make_output` in addition being a super trait to `Pin`

phil-levis replied:
Configure allows you to change how the pin operates; it's broken out from Input and Output so that you can give access to read or write a pin without also granting the ability to change its operating mode. The Pin trait is more full access to a Pin: you can configure it, read it, and write it.

alevy replied:
It's the old `PinCtl`. I think it makes a lot of sense as a trait, I'm a bit confused by some of the specific methods or their signatures in it (we can discuss during the call?).

@niklasad1 totally agree though, the final version (before merging) should (and will) definitely have much more documentation.

niklasad1 commented on kernel/src/hil/gpio.rs:
```diff
@@ -1,67 +1,289 @@
+/// `gpio::ClientWithValue::fired`.
+pub struct InterruptValueWrapper {
+    value: Cell<u32>,
+    client: OptionalCell<&'static ClientWithValue>,
```
Would be nice to use `dyn` in all use of `trait objects` because trait objects without explicit `dyn` is deprecated in later rust versions 

For example `client: OptionalCell<&'static dyn ClientWithValue> `

phil-levis replied:
This seems like something we should do consistently over the entire repository. It seems like a good thing to propose for 1.4.

niklasad1 commented on chips/stm32f4xx/src/spi.rs:
```diff
@@ -5,7 +5,7 @@ use kernel::common::cells::OptionalCell;
 use kernel::common::StaticRef;
 use kernel::hil;
-use kernel::hil::gpio::Pin;
+use kernel::hil::gpio::{Output};
```
```suggestion
use kernel::hil::gpio::Output;
```

phil-levis replied:
OK

niklasad1 commented on chips/stm32f4xx/src/gpio.rs:
```diff
@@ -1026,16 +1139,18 @@ impl hil::gpio::Pin for Pin<'a> {
+        unsafe {
+            let val = self.exti_lineid.map(|lineid| {
+                let l = lineid.clone();
+                exti::EXTI.is_pending(l)
```
`lineid` is copy dereference instead.

Also you could use map_or instead of `map + unwrap_or`

phil-levis replied:
Yes, much better!

niklasad1 commented on capsules/src/mcp230xx.rs:
```diff
@@ -528,9 +524,13 @@ impl hil::i2c::I2CClient for MCP230xx<'a> {
-    fn fired(&self, bank_number: usize) {
+impl gpio::ClientWithValue for MCP230xx<'a> {
+    fn fired(&self, value: u32) {
+        if value != 0 && value != 1 {
```
could be simplified to `if value < 2 { break; }`

phil-levis replied:
Better inequality but I prefer return because it's more explicit that it means "stop now."

niklasad1 commented on boards/nucleo_f429zi/src/main.rs:
```diff
@@ -263,36 +267,53 @@ pub unsafe fn reset_handler() {
+        [kernel::hil::gpio::InterruptValueWrapper::new()]
+    );
+    
+    for i in 0..NUM_BUTTONS {
```
prefer `iterators` over `ranges `

for example do:

```rust
    for (&pin, value) in button_pins.iter().zip(values.iter()) {
        pin.set_client(value);
        value.set_source(pin);
    }
```

niklasad1 commented on boards/nucleo_f429zi/src/main.rs:
```diff
@@ -263,36 +267,53 @@ pub unsafe fn reset_handler() {
-    for &(btn, _) in button_pins.iter() {
-        btn.set_client(button);
+
+    for i in 0..NUM_BUTTONS {
```
prefer iterators over ranges

phil-levis replied:
Generally, yes, but in this case I was trying to keep a parallel structure to the earlier initialization code that iterated over both values and button_pins.

ppannuto commented:
I fixed up the board file for hail (tested on HW with buttons app) and then ported over the cc26x2 (compile-tested only).

There is one minor hangup for the cc26x2 w.r.t. this interface: it's not clear that the chip supports checking whether there are any interrupts pending ([Reference Manual, §13](http://www.ti.com/lit/ug/swcu185c/swcu185c.pdf)). I don't want to claim that strongly, but a method for checking pending status wasn't immediately obvious on first read. For now [I left `is_pending` as unimplemented](https://github.com/tock/tock/pull/1297/commits/135fc23e3c3384b22d2e130e64544764fbc604b9#diff-0d6cbeeb5e5eb9b8b368ddedffffdd7fR397), but I'm not really sure that's the best long-term solution.

ppannuto commented:
Summary of current status (I believe, please edit this comment if corrections are needed):

Open HIL questions:

 - [ ] [opaque configuration](https://github.com/tock/tock/issues/1233#issuecomment-455283781)? I still think this could be useful for transparent multiplexing, but could be disabused of that notion
 - [ ] Remove `is_pending` from HIL? This was not in the proposal in #1233, nor is it currently used by Tock. As the cc26xx doesn't have an obvious path to support this and it doesn't seem to have demand for its use, the suggestion is to drop this method.

Open implementation work:

 - [ ] nrf5x <-- @alevy

phil-levis commented:
Great -- I will go through the comments and respond. Sorry for the delay.

phil-levis commented:
I think `is_pending` is necessary; if you want to check or poll an interrupt line, you need it. If you're unable to check if an interrupt is pending then you can't have multiple interrupts invoke the same handler (since you don't know which one triggered your handler).

alevy commented:
> If you're unable to check if an interrupt is pending then you can't have multiple interrupts invoke the same handler (since you don't know which one triggered your handler).

This is supposedly the reason for `InterruptValuePin` and `ClientWithValue`, isn't it?

alevy commented:
NRF52 is working, though I also made a change to the `InterruptValueWrapper` and propagated the changes to all the boards (which should be much much cleaner now, but it's possible I made mistakes and am not able to test all of them).

All but one of the two nucleo boards (which hasn't been ported) is at least compiling now, and the format check is passing.

phil-levis commented:
> This is supposedly the reason for InterruptValuePin and ClientWithValue, isn't it?

Yes, which requires a word of state and extra code. A much more common kernel design pattern is

```handler() {
  if a is pending {A();}
  if b is pending {B();}
```

I kept InterruptValuePin primarily for userspace; the above pattern is good within the kernel, but is not great across the user/kernel boundary. It would require, in part, software-based pending bits for every interrupt.

alevy commented:
> A much more common kernel design pattern is

You mean a common design pattern in kernels generally, not in Tock specifically, right? In Tock there are currently no uses of the `is_pending` currently (though maybe you plan on having it in the Titan repo).

phil-levis commented:
> You mean a common design pattern in kernels generally, not in Tock specifically, right? In Tock there are currently no uses of the is_pending currently (though maybe you plan on having it in the Titan repo).

Yes.

It's very important to be able to check for pending interrupts. `disable_interrupt` is quite dangerous because if you don't have `is_pending`, you could lose/delay interrupts for a long time . E.g., "hey, a timeout occurred, let's turn off the radio and disable interrupts on those pins so we can go into an ultra-low-power MUCH state, oops, a packet arrived while we were doing that."

ppannuto commented:
Upon a second digging into of this datasheet, I think this might be the register to read to implement `is_pending` for the cc26x2:

<img width="756" alt="image" src="https://user-images.githubusercontent.com/339422/61554366-a15e7980-aa11-11e9-8a40-a2784f4aa336.png">

phil-levis commented:
Based on the discussions, I added a ConfigurationInputOutput trait and changed `Configuration::Unknown` to `Configuration::Other`. I added a bunch of comments and defined what the `Configuration` states mean.

ppannuto commented:
We talked about this a bit on the call, I regret not writing down thoughts earlier, but I think the major bullet points:

 - If it is necessary to check the return to ensure that the requested configuration was actually set, then we should a return type that more explicitly indicates that (e.g. `Result<Configuration, Configuration>` (which returns the actual HW configuration in either case, and `Ok` only if it changed to what was requested)
 - The `make_*` family of functions feels awkward. They all have to be implemented since they are part of one interface, so there's no indication from these broken out functions of what can or can't be "made" for this pin. This also means that there is a configuration enum that is used only as a return value to see what was actually set. Especially now that `inout` is in its own interface, I think the weak consensus was that the motivation for a `make_*` family of functions has sort of gone away, and that a `set/get_configuration` interface may make more sense

phil-levis commented:
I kinda feel like we're going in circles.

First, there Brad's suggestion that the ability to make something simultaneously an input and output be type checked; i.e., if you depend on this functionality, then there's a specific trait you use so that if it doesn't exist it's a compile-time error. I agree with his sentiment on this -- the more we can check at compile-time, the better.

But now the suggestion is to push checks back to runtime. I.e., `get_configuration` and `set_configuration`. These traits have been around for almost 6 months, and now we seem to be in a perturbation, rather than refinement, state.

ppannuto commented:
I think my notes may have been a little too brief. IIRC, consensus was that we are happy with leaving the `InOut` in a separate trait. The weird bit was the asymmetry of the setters being a family of `make_*` functions, and the getter being a generic `configuration` function.

However. Upon a fresh read, what's there is symmetric (`make_input`/`is_input`), and the generic `configuration` is really more in service of the weird cases that may have been set outside of this HIL. I think I'm actually happy with what is there. If anyone else has a different memory of what we were talking about on the call, holler, but I've convinced myself I'm happy with what is here.

phil-levis commented:
The intention with the make_ and is_ calls is that they are the primary way that a client will interact with the driver. This allows them to behave somewhat independently of whether the underlying driver supports simultaneous input and output. For example, suppose I have a capsule that's driving an output pin (and using the Output trait). I can, if I want, have another capsule monitor this pin by making it an input (and using the Input trait). For chips which support simultaneous input/output, this will work. For chips that don't, it won't. But, it means that I don't have to change the capsule that makes the pin an output (e.g., by having it make the pin an InputOutput). The point is that for some chips, whether a pin is an input or an output is independent, and so we should treat it as such.

bradjc commented on kernel/src/hil/gpio.rs:
```diff
@@ -1,67 +1,326 @@
+    /// Disable the pin and put it into its lowest power state.
+    /// Re-enabling the pin requires reconfiguring it (state of
+    /// its enabled configuration is not stored).
+    fn low_power(&self);
```
What about making this (and I can push the commit) `deactivate_to_low_power()` (or something)? I think that seeing a call like `my_pin.low_power()` won't be intuitive that that actually stops the pin from working unless you have studied this HIL. Also, it seems like you would always want to call `my_pin.low_power()` because who would want a high power pin? And since the function doesn't guarantee to actually _disable_ the pin, "deactivate" seemed like an appropriate verb to signal that things likely won't work after calling this function.

phil-levis replied:
Yeah, that makes sense. Or `put_in_low_power()`.

bradjc commented:
BTW, I'm supportive of just merging this and testing for 1.4 rather than testing twice.
